### DISCUSSION TITLE
printf chewing up a lot of ram, alternative to F("  ")?

### DISCUSSION THREAD

**Discussion by microPaul** [SpenceKonde/DxCore#431] [Q&A]
I like to use printf statements (in spite of the shortcomings).  I used a lot of them for a project (help files).  Of course I was reminded that strings, although compiled to flash, do get put into ram, chewing up a lot of ram.  So I wrapped all those essentially constant strings in F("  ") and that solved the ram use problem.

`Serial.printf(F("there are %d sides to a square\n"), 4);`

But I wonder if there's some sort of global compiler directive that will cause all strings  (those that are essentially constants, determined at compile time) to only be in flash and not get copied for ram?  I mean the F("  ") works with Serial.printf, but I'd rather not include F("  ") in every printf statement, if it can be avoided.

Any suggestions or guidance as to a possible work-around?

**Comment by SpenceKonde** [comment]
Nope, there is not. F() macro is an Arduino thing. There's no compiler directive because it's outside the realm of the compiler. GCC does not understand the multiple address spaces very well. I mean just look at the assembly listings. You see locations in program space listed as offsets from landmarks in the data space and vise versa. 

I would suggest regex find `([^F][^\(])"(.*)?"` replace `\1F("\2")`
matches anything that starts with something other than a capital F, and anything other than a open paren as the second character, followed by a quote, and the shortest string of text without a linebreak in it which is followed by the closing quote, then replaces it with the first two letters (so it doesn't trash the letters you checked weren't already `F(` followed by a " and then the text before the closing quote. 

There is no way I could maintain these cores withoutextensive use of regexes. Regex's are IMO one of the most useful "languages" to know, because of these extreme versatility. Processing data to reformat it? Regex. Making .hex files more human readable? Regexes. Making mass changes to source code? regex. 
Yes, the syntax is outlandish, but the power is too great to ignore.

**Comment by SpenceKonde** [comment]
Frankly, F() already Is a workaround. You don't wanna know what the non-arduino way to stuff strings into flash is.

**Comment by microPaul** [comment]
Thank you Spence.  I appreciate your advice and expertise.   I used to have some rudimentary knowledge of regular expressions when I did  a couple of small projects in Perl, about 8 years ago.  I still have "Mastering Regular Expressions" on the bookshelf right above my desk.  Sounds like it's time to do some new learning and re-learning.

thanks,

Paul

**Reply by SpenceKonde** [reply]
Get a text editor like sublime text that lets you do regex search/replace from within same application imo. And like - I started with really basic stuff. It's incredibly helpful to have the find and replace fields staring at you, each of which has it's own undo buffer as does the main document, etc. So when you screw up with a regex and turn everything into garbage, you can just click the main text part of the window ctrl+z and it's undone, and you fix the regex. That editor also lets you "find all" occurrences that match something, which makes them all act selected, so you can cut/copy/paste/delete them all at once. How do you think I built core_devices.h?

**Comment by microPaul** [comment]
Terrific!  Thanks Spence!
